### DISCUSSION TITLE
Cannot upload attachments using the Bitwarden CLI

### DISCUSSION THREAD

**Discussion by aligator** [dani-garcia/vaultwarden#3074] [Q&A]
### Cannot upload attachments using the Bitwarden CLI

When it uploads attachments using the Bitwarden CLI, I always get:
`bw create attachment --file "/tmp/lol.txt" --itemid ddfd114b-ca18-45bc-b8e4-42da43a71bde`

```
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <title>400 Bad Request</title>
</head>
<body align="center">
    <div role="main" align="center">
        <h1>400: Bad Request</h1>
        <p>The request could not be understood by the server due to malformed syntax.</p>
        <hr />
    </div>
    <div role="contentinfo" align="center">
        <small>Rocket</small>
    </div>
</body>
</html>
```
and on the Vaultwarden server-side I get:
```
vaultwarden    | [2022-12-30 13:49:36.453][_][WARN] Data guard `Form < UploadData < '_ > >` failed: Errors([Error { name: Some("data"), value: None, kind: Io(Custom { kind: Other, error: field "data" received with incomplete data }), entity: Form }]).
vaultwarden    | [2022-12-30 13:49:36.453][_][WARN] No 400 catcher registered. Using Rocket default.
```

### Your environment (Generated via diagnostics page)
* Vaultwarden version: v1.27.0
* Web-vault version: v2022.12.0
* Running within Docker: true (Base: Debian)
* Environment settings overridden: true
* Uses a reverse proxy: true
* IP Header check: true (X-Forwarded-For)
* Internet access: true
* Internet access via a proxy: false
* DNS Check: true
* Time Check: true
* Domain Configuration Check: true
* HTTPS Check: true
* Database type: SQLite
* Database version: 3.39.2
* Clients used: 
* Reverse proxy and version: 
* Other relevant information: 

### Config (Generated via diagnostics page)
<details><summary>Show Running Config</summary>

**Environment settings which are overridden:** ADMIN_TOKEN

```json
{
  "_duo_akey": null,
  "_enable_duo": false,
  "_enable_email_2fa": true,
  "_enable_smtp": true,
  "_enable_yubico": true,
  "_icon_service_csp": "",
  "_icon_service_url": "",
  "_ip_header_enabled": true,
  "_smtp_img_src": "cid:",
  "admin_ratelimit_max_burst": 3,
  "admin_ratelimit_seconds": 300,
  "admin_token": "***",
  "allowed_iframe_ancestors": "",
  "attachments_folder": "data/attachments",
  "authenticator_disable_time_drift": false,
  "data_folder": "data",
  "database_conn_init": "",
  "database_max_conns": 10,
  "database_timeout": 30,
  "database_url": "***************",
  "db_connection_retries": 15,
  "disable_2fa_remember": false,
  "disable_admin_token": false,
  "disable_icon_download": false,
  "domain": "*****://*****************",
  "domain_origin": "*****://*****************",
  "domain_path": "",
  "domain_set": true,
  "duo_host": null,
  "duo_ikey": null,
  "duo_skey": null,
  "email_attempts_limit": 3,
  "email_expiration_time": 600,
  "email_token_size": 6,
  "emergency_access_allowed": true,
  "emergency_notification_reminder_schedule": "0 3 * * * *",
  "emergency_request_timeout_schedule": "0 7 * * * *",
  "enable_db_wal": true,
  "event_cleanup_schedule": "0 10 0 * * *",
  "events_days_retain": null,
  "extended_logging": true,
  "helo_name": null,
  "hibp_api_key": null,
  "icon_blacklist_non_global_ips": true,
  "icon_blacklist_regex": null,
  "icon_cache_folder": "data/icon_cache",
  "icon_cache_negttl": 259200,
  "icon_cache_ttl": 2592000,
  "icon_download_timeout": 10,
  "icon_redirect_code": 302,
  "icon_service": "internal",
  "incomplete_2fa_schedule": "30 * * * * *",
  "incomplete_2fa_time_limit": 3,
  "invitation_expiration_hours": 120,
  "invitation_org_name": "Vaultwarden",
  "invitations_allowed": true,
  "ip_header": "X-Forwarded-For",
  "job_poll_interval_ms": 30000,
  "log_file": null,
  "log_level": "Info",
  "log_timestamp_format": "%Y-%m-%d %H:%M:%S.%3f",
  "login_ratelimit_max_burst": 10,
  "login_ratelimit_seconds": 60,
  "org_attachment_limit": null,
  "org_creation_users": "***************",
  "org_events_enabled": false,
  "org_groups_enabled": false,
  "password_hints_allowed": true,
  "password_iterations": 100000,
  "reload_templates": false,
  "require_device_email": false,
  "rsa_key_filename": "data/rsa_key",
  "send_purge_schedule": "0 5 * * * *",
  "sends_allowed": true,
  "sends_folder": "data/sends",
  "show_password_hint": false,
  "signups_allowed": false,
  "signups_domains_whitelist": "",
  "signups_verify": true,
  "signups_verify_resend_limit": 6,
  "signups_verify_resend_time": 3600,
  "smtp_accept_invalid_certs": false,
  "smtp_accept_invalid_hostnames": false,
  "smtp_auth_mechanism": null,
  "smtp_debug": false,
  "smtp_embed_images": true,
  "smtp_explicit_tls": null,
  "smtp_from": "********************",
  "smtp_from_name": "Vaultwarden",
  "smtp_host": "*****************",
  "smtp_password": "***",
  "smtp_port": 587,
  "smtp_security": "starttls",
  "smtp_ssl": null,
  "smtp_timeout": 15,
  "smtp_username": "********************",
  "templates_folder": "data/templates",
  "tmp_folder": "data/tmp",
  "trash_auto_delete_days": 365,
  "trash_purge_schedule": "0 5 0 * * *",
  "use_syslog": false,
  "user_attachment_limit": 5000000,
  "web_vault_enabled": true,
  "web_vault_folder": "web-vault/",
  "websocket_address": "0.0.0.0",
  "websocket_enabled": true,
  "websocket_port": 3012,
  "yubico_client_id": null,
  "yubico_secret_key": null,
  "yubico_server": null
}
```
</details>

* Install method: Docker

* Clients used: Bitwarden-CLI

### Steps to reproduce
Nothing special, just use the bw cli comand from above

**Comment by BlackDex** [comment]
Works fine for me.
Check your reverse proxy config if it isn't using some security extensions or settings like ModSecurity or whatever.
Also check the max body size which is allowed, those could cause issues.

I just tested it using:

```bash
bw create attachment --file favicon-16x16.png --itemid a0f5ec26-9d2c-4f97-b4f9-ac06015bd425
```

And that works ok for me.

**Reply by aligator** [reply]
thanks for the fast answer, so it must be a configuration error.
I will look into it ... :-)

**Reply by aligator** [reply]
hm, interestingly only the cli does not work. Everywhere else the file upload works perfectly (web, app, desktop-app, ...).
Does it do the upload differently in the cli?

However, If that's the only problem, all is fine, as I won't need the upload with the cli that much^^

**Reply by BlackDex** [reply]
Which version of the cli do you use? And where does the cli binary come from? What if you download a pre-build one from the bitwarden/clients repo?

**Reply by aligator** [reply]
It is from the arch repository 2022.11.0 https://archlinux.org/packages/community/any/bitwarden-cli/

Now that is interesting:
 (/tmp/bw is downloaded from https://github.com/bitwarden/clients/releases/tag/cli-v2022.11.0)
```
➜  ~ bw --version
2022.11.0
➜  ~ /tmp/bw --version
2022.11.0
➜  ~ /tmp/bw create attachment --file /tmp/vaultwarden-favicon.png --itemid ddfd114b-ca18-45bc-b8e4-42da43a71bde
{"object":"item","id":"ddfd114b-ca18-45bc-b8e4-42da43a71bde",{...CENSORED...}"}],"revisionDate":"2022-12-30T13:43:10.951Z","creationDate":"2022-12-30T13:43:10.951Z","deletedDate":null}
➜  ~ bw create attachment --file /tmp/vaultwarden-favicon.png --itemid ddfd114b-ca18-45bc-b8e4-42da43a71bde
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <title>400 Bad Request</title>
</head>
<body align="center">
    <div role="main" align="center">
        <h1>400: Bad Request</h1>
        <p>The request could not be understood by the server due to malformed syntax.</p>
        <hr />
    </div>
    <div role="contentinfo" align="center">
        <small>Rocket</small>
    </div>
</body>
</html>
```

So what's the difference?
The version from arch is built with this: https://github.com/archlinux/svntogit-community/blob/packages/bitwarden-cli/trunk/PKGBUILD

I am just building myself from that PKGBUILD to see if it makes any difference...

**Reply by aligator** [reply]
With the self-built version (`makepkg -si`) it also doesn't work.
Any idea what's the difference to the official build?

**Reply by BlackDex** [reply]
Yes, the NPM/Node version is different.
I have reported this a while back at Arch, and they have fixed it.
If you have a newer version of node, then it will not work.

**Reply by aligator** [reply]
https://bugs.archlinux.org/task/76967

**Reply by BlackDex** [reply]
But, which node version are you using to build it?

**Reply by BlackDex** [reply]
Because, why do you build it your self instead of using the community provided package?

**Reply by BlackDex** [reply]
Ok, it looks like you use NodeJS 18 aka nodejs-lts-hydrogen.
That isn't compatible, you need 16 aka nodejs-lts-gallium

**Reply by aligator** [reply]
oh, now I understand, I always thought I use the default arch node, but forgot that I have nvm installed^^ Also I was not aware that the release binary from github uses a packaged node.
It is all a bit confusing. (I hate the version management with node and java, always the same problems^^)

I wonder if there would be any way to avoid that someone else has the same problems, as it is very non-obvious that  bw (on arch) breaks with a newer node in a way which just tells nothing about the real cause...

However many thanks for your help and time.

**Reply by aligator** [reply]
> Because, why do you build it your self instead of using the community provided package?   

That was just for testing, but the comunity package also doesn't work if I have a newer node in the PATH.

**Reply by BlackDex** [reply]
The community packages work fine on my system. The reason it doesn't work for you is probably because you have Node v18.

Using Node v16 has no issues at all.
I also build it my self (Using the correct node version), and that also works just fine.

You also mention, a **newer** version of node, but that is just the issue. It needs an **older** version.

**Reply by aligator** [reply]
Yes, but it sadly fails in a way, where I would never think of a wrong node version... (I even didn't know a view hours before that it is written with node^^)

But I think we cannot do anything about it for now and just hope that no one else is  that dumb to use nvm with non-system-default like I am.

**Reply by BlackDex** [reply]
Hehe, i think that was discussed in the issue i posted.
https://bugs.archlinux.org/task/74929

**Reply by aligator** [reply]
The only way I could think of is adding a Wrapper-script which first checks the node version and prints a warning if it is not supported (at run time).

Because at build time, there is nothing you can do, as it fails at run time, even when the node version at build time was correct.

Nvm would not help you in the PKGBUILD, because 1. it is from AUR and not supported in community and 2. it would only help for the build time requirement, not for the run time.

**Reply by aligator** [reply]
Also if I understand it correctly, even your bugreport was not the first issue due to different node versions in arch and the official package.

BTW, how is the official package packed? It seems to include the node version it needs. wouldn't that be also possible in the PKGBUILD?

**Reply by aligator** [reply]
ah, ok it uses https://www.npmjs.com/package/pkg
That way its just independent from the installed node (at runtime).

**Reply by BlackDex** [reply]
I would suggest to check:
https://github.com/bitwarden/clients/blob/3d9e633233a5309a88c2efff282c00d120a09a7c/.github/workflows/build-cli.yml
https://github.com/bitwarden/clients/blob/3d9e633233a5309a88c2efff282c00d120a09a7c/.github/workflows/release-cli.yml
